libevm/triedb/firewood/firewood.go

```diff
@@ -0,0 +1,518 @@
+// or (at your option) any later version.
+//
+// The libevm additions are distributed in the hope that they will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU Lesser
+// General Public License for more details.
+//
+// You should have received a copy of the GNU Lesser General Public License
+// along with the go-ethereum library. If not, see
+// <http://www.gnu.org/licenses/>.
+
+// The firewood package provides a [triedb.DBOverride] backed by [Firewood].
+//
+// [Firewood]: https://github.com/ava-labs/firewood
+package firewood
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"os"
+	"path/filepath"
+	"sync"
+
+	"github.com/ava-labs/firewood-go-ethhash/ffi"
+
+	"github.com/ava-labs/libevm/common"
+	"github.com/ava-labs/libevm/core/rawdb"
+	"github.com/ava-labs/libevm/core/types"
+	"github.com/ava-labs/libevm/ethdb"
+	"github.com/ava-labs/libevm/libevm/stateconf"
+	"github.com/ava-labs/libevm/log"
+	"github.com/ava-labs/libevm/trie/trienode"
+	"github.com/ava-labs/libevm/trie/triestate"
+	"github.com/ava-labs/libevm/triedb"
+	"github.com/ava-labs/libevm/triedb/database"
+)
+
+const (
+	dbFileName  = "firewood.db"
+	logFileName = "firewood.log"
+)
+
+var _ triedb.DBOverride = (*Database)(nil)
+
+type Database struct {
+	// The underlying Firewood database, used for storing proposals and revisions.
+	// This must be exported so other packages (e.g. state sync) can access firewood-specific methods.
+	Firewood *ffi.Database
+
+	// proposalLock protects the proposal map and tree during updates.
+	proposalLock sync.RWMutex
+	// proposalMap provides O(1) access by state root to all proposals stored in the proposalTree
+	proposalMap map[common.Hash][]*proposal
+	// The proposal tree tracks the structure of the current proposals, and which proposals are children of which.
+	// This is used to ensure that we can dereference proposals correctly and commit the correct ones
+	// in the case of duplicate state roots.
+	// The root of the tree is stored here, and represents the top-most layer on disk.
+	proposalTree *proposal
+
+	// possibleProposals temporarily holds proposals created during a trie update.
+	// This is cleared after the update is complete and the proposals have been sent to the database.
+	possibleProposals []*proposal
+}
+
+type Config struct {
+	ChainDir             string
+	CleanCacheSize       int  // Size of the clean cache in bytes
+	FreeListCacheEntries uint // Number of free list entries to cache
+	Revisions            uint
+	ReadCacheStrategy    ffi.CacheStrategy
+}
+
+// Note that `FilePath` is not specified, and must always be set by the user.
+var Defaults = Config{
+	CleanCacheSize:       1024 * 1024, // 1MB
+	FreeListCacheEntries: 40_000,
+	Revisions:            100,
+	ReadCacheStrategy:    ffi.CacheAllReads,
+}
+
+func (c Config) BackendConstructor(ethdb.Database) triedb.DBOverride {
```

Comment:
```suggestion
var _ triedb.DBConstructor = (Config{}).BackendConstructor

// BackendConstructor is a [triedb.DBConstructor], expected to be used in a [triedb.Config].
func (c Config) BackendConstructor(ethdb.Database) triedb.DBOverride {
```

The method comment is for documentation servers while the `var _` locks in the function signature. This is useful here (as proof of correctness) and in case we change `DBConstructor` as we then know what we've broken and need to fix.